core/storage/trie_pruner/impl/trie_pruner_impl.cpp

```diff
@@ -110,86 +128,105 @@ namespace kagome::storage::trie_pruner {
-    }
-
-    trie::MerkleValue getMerkleValue(const trie::DummyNode &node) const {
-      return node.db_key;
-    }
+  bool TriePrunerImpl::start() {
+    prune_thread_handler_->execute(
+        [self = shared_from_this()]() { self->pruneQueuedStates(); });
+    return true;
+  }
 
-    std::optional<common::Hash256> getValueHash(const trie::TrieNode &node,
-                                                trie::StateVersion version) {
-      if (node.getValue().is_some()) {
-        if (node.getValue().hash.has_value()) {
-          return *node.getValue().hash;
-        }
-        if (codec.shouldBeHashed(node.getValue(), version)) {
-          return codec.hash256(*node.getValue().value);
-        }
+  void TriePrunerImpl::pruneQueuedStates() {
+    PendingPrune prune;
+    while (prune_queue_.pop(prune)) {
+      prune_queue_length_--;
+      switch (prune.reason) {
+        case PruneReason::Finalized:
+          if (auto res = pruneFinalized(prune.root, prune.block_info);
+              res.has_error()) {
+            SL_WARN(logger_,
+                    "Failed to prune finalized block {}: {}",
+                    prune.block_info,
+                    res.error());
+          }
+          break;
+        case PruneReason::Discarded:
+
+          if (auto res = pruneDiscarded(prune.root, prune.block_info);
+              res.has_error()) {
+            SL_WARN(logger_,
+                    "Failed to prune discarded block {}: {}",
+                    prune.block_info,
+                    res.error());
+          }
+          break;
+      }
+      SL_DEBUG(logger_, "Prune queue size: {}", prune_queue_length_);
+      // To let new blocks pass through, otherwise new blocks wait too long
+      // for pruner mutex.
+      // During normal sync (not catch-up) though this queue may pile up too
+      // quickly without this limit.
+      if (prune_queue_length_ < 1000) {
+        break;
       }
-      return std::nullopt;
-    }
-
-    outcome::result<trie::MerkleValue> getMerkleValue(
-        const trie::TrieNode &node, trie::StateVersion version) {
-      encode_called++;
-      OUTCOME_TRY(
-          merkle_value,
-          codec.merkleValue(
-              node,
-              version,
-              [this](trie::Codec::Visitee visitee) -> outcome::result<void> {
-                if (auto child_data =
-                        std::get_if<trie::Codec::ChildData>(&visitee);
-                    child_data != nullptr) {
-                  return visitChild(child_data->child,
-                                    child_data->merkle_value);
-                }
-                return outcome::success();
-              }));
-      return merkle_value;
     }
+    prune_thread_handler_->withIoContext(
+        [self = shared_from_this()](auto &io_ctx) {
+          // to let new blocks pass through, otherwise new blocks wait too long
+          // for pruner mutex
+          auto timer = std::make_shared<boost::asio::steady_timer>(
```

Comment:
Should we create new timer each time or reuse existing?

Reply:
I think it's a bit more convenient code-wise to just create a new one, so since this operation is performed rarely, I'd leave it as is.